syntax = "proto3";
package protocol;

import "common.proto";

message Signer {
	enum Limit{
		SIGNER_NONE = 0;
		SIGNER = 100;
	};
	string address = 1;
	int64 weight = 2;
}

message ControlledArea{
	enum TRUST_FLAG {
		UNCERT        = 0;
		TRUSTED       = 1;
		UNTRUSTED     = 2;
	};
	
	enum STATUS_FALG {
		ENABLED        = 0;
		DISABLED       = 1;
	};
	
    TRUST_FLAG trusted = 1;
	STATUS_FALG status = 2;
	repeated KeyPair attributes = 3 ;
}

message OperationTypeThreshold{
	Operation.Type type = 1;
	int64 threshold = 2;
}

message Contract{
    enum ContractType{
		JAVASCRIPT = 0;
		EVM = 1;
		SYSTEM = 2;
	}
	ContractType type = 1;
	string payload = 2;
	int64 version = 3;
}

message OperationPayCoin{
	string dest_address = 1;
	int64 amount = 2;
	string input = 3;
	bytes raw_dest_address = 4;
}

message OperationSetSignerWeight{
	int64 master_weight = 1; //required, [-1,MAX(UINT32)] -1: indicates no setting
	repeated Signer signers = 2; //address:weight, if weight is 0, it indicates deleting this signer
}

message OperationLog{
	string topic = 1;
	repeated string datas = 2;
    repeated string topics = 3; //for evm
}

message OperationSetPrivilege{
	string master_weight = 1;
	repeated Signer signers = 2;
	string tx_threshold = 3;
	repeated OperationTypeThreshold type_thresholds = 4;
}

message OperationAuthorizeTransfer{
    string dest_address = 1;
    bool   auth = 2;
}

message AccountPrivilege {
	int64 master_weight = 1;
	repeated Signer signers = 2;
	AccountThreshold thresholds = 3;
}

message AccountThreshold{
	int64 tx_threshold = 1; //required, [-1,MAX(INT64)] -1: indicates no setting
	repeated OperationTypeThreshold type_thresholds = 2;
}

message OperationCreateAccount{
	string dest_address = 1;
	Contract contract = 2;
	AccountPrivilege priv = 3;
	repeated KeyPair metadatas = 4;	
	int64	init_balance = 5;
	string  init_input = 6;
	string dest_public_key = 7;
	bytes raw_dest_address = 8;
    bytes salt_dest_address = 9;
}

message OperationSetMetadata{
	string	key = 1;  
	string  value = 2;
	int64 	version = 3; //Optional. The version number of metadata. If set, it must be the same as the existing metadata_version of the account.
	bool    delete_flag = 4;
}

message OperationUpgradeContract {
	string dest_address = 1;
	string payload = 2;
	int64  type = 3;
}

message OperationSetControlledArea {
	string dest_address = 1;
	bool s_trusted = 2;
	bool s_status = 3;
    ControlledArea.TRUST_FLAG trusted = 4;
	ControlledArea.STATUS_FALG status = 5;
	repeated KeyPair attributes = 6; 
}

message Operation {
	enum Type {
		UNKNOWN = 0;
		CREATE_ACCOUNT 			= 1;
		SET_METADATA			= 4;
		SET_SIGNER_WEIGHT		= 5;
		SET_THRESHOLD			= 6;
		PAY_COIN                = 7;
		LOG						= 8;
		SET_PRIVILEGE			= 9;
		UPGRADE_CONTRACT		= 10;
		SET_CONTROLLED_AREA     = 11;
        AUTHORIZE_TRANSFER      = 12;
	};
	Type type = 1;
	string source_address = 2;
	bytes metadata	= 3;

	OperationCreateAccount		create_account 	   = 4;
	OperationSetMetadata		set_metadata	   = 7;
	OperationSetSignerWeight	set_signer_weight  = 8;
	OperationSetThreshold		set_threshold 	   = 9;
	OperationPayCoin			pay_coin           = 10;
	OperationLog				log				   = 11;
	OperationSetPrivilege		set_privilege	   = 12;
	OperationUpgradeContract	upgrade_contract   = 13;
	OperationSetControlledArea	set_controlled_area   = 14;
    OperationAuthorizeTransfer  authorize_transfer  = 15;
	bytes raw_source_address = 20;
}

message OperationSetThreshold{
	int64 tx_threshold = 1;
	repeated OperationTypeThreshold type_thresholds = 4; //type:threshold, threshold:0 indicates deleting this type
}

message Transaction {
	enum Limit{
		UNKNOWN = 0;
		OPERATIONS = 1000;
	};
    enum TxType{
        INCREASE_NONCE   = 0;
        RANDOM_NONCE   = 1;
    };
	string source_address = 1;
	int64 nonce = 2;
	string expr_condition = 3;
	repeated Operation operations = 4;
	bytes metadata = 5;
	int64  fee_limit = 6;
	int64  gas_price =7;
	int64 ceil_ledger_seq = 8;
	int64 chain_id = 9;
	string address_prefix = 10; //it represent the address is raw
	bytes raw_source_address = 11;
	int32 domain_id = 12;      // 废弃，暂不删除
	TxType nonce_type = 13;
	int64 max_ledger_seq = 14;
}